zokrates_core/src/ir/expression.rs

```diff
@@ -184,23 +210,27 @@ mod tests {
             let a: LinComb<FieldPrime> = FlatVariable::new(42).into();
             let b: LinComb<FieldPrime> = FlatVariable::new(42).into();
             let c = a + b.clone();
-            let mut expected_map = BTreeMap::new();
-            expected_map.insert(FlatVariable::new(42), FieldPrime::from(2));
-            assert_eq!(c, LinComb(expected_map));
+
+            let expected_vec = vec![
+                (FlatVariable::new(42), FieldPrime::from(1)),
+                (FlatVariable::new(42), FieldPrime::from(2)),
+
+            ];
+            assert_eq!(c, LinComb(expected_vec));
         }
         #[test]
         fn sub() {
             let a: LinComb<FieldPrime> = FlatVariable::new(42).into();
             let b: LinComb<FieldPrime> = FlatVariable::new(42).into();
             let c = a - b.clone();
-            assert_eq!(c, LinComb::zero());
+            assert_eq!(LinComb::zero(), c);
         }
 
         #[test]
         fn display() {
             let a: LinComb<FieldPrime> =
                 LinComb::from(FlatVariable::new(42)) + LinComb::summand(3, FlatVariable::new(21));
-            assert_eq!(&a.to_string(), "3 * _21 + 1 * _42");
+            assert_eq!(&a.to_string(), "1 * _42 + 3 * _21");
```

Comment:
It would make sense for `fmt::Display` to reduce to canonical before printing, so that this test ideally would not change?

Reply:
Can you please clarify what a canonical representation is? I probably should implement a method `as_canonical() -> Self` for a lincomb and use that in `fmt::Display` and in communication with backend.

Reply:
Yes maybe we need to define some types here for it to be clearer. By canonical I mean BTreeMap based, where you have a single, deterministic way to represent a given linear combination (BTreeMap adds the ordered aspect which you don't get from HashMap)

Reply:
Can we just sort the underlying `Vec` by variable ID?

Something, like:
```rust
pub fn as_canonical(&self) -> Self {
  let mut sorted = self.0.clone();
  sorted.sort_by_key(|(id, _)| id);

  LinComb(sorted)
}
```

Reply:
Sorting is one thing, but we mostly need each variable to appear only once in each linear combination, which the vector does not enforce. Hence the BTreeMap imo.